ocaml/xapi/nm.ml

```diff
@@ -38,55 +38,73 @@ let determine_mtu pif_rc net_rc =
     let value = List.assoc "mtu" pif_rc.API.pIF_other_config in
     try int_of_string value
     with _ ->
-      debug "Invalid value for mtu = %s" value ;
+      warn "Invalid value for other-config:mtu = %s" value ;
       mtu
   ) else
     mtu
 
 let determine_ethtool_settings properties oc =
   let proc key =
-    if List.mem_assoc ("ethtool-" ^ key) oc then
-      let value = List.assoc ("ethtool-" ^ key) oc in
-      if value = "true" || value = "on" then
+    match
+      (List.assoc_opt ("ethtool-" ^ key) oc, List.assoc_opt key properties)
+    with
+    | Some ("true" | "on"), _ ->
         [(key, "on")]
-      else if value = "false" || value = "off" then
+    | Some ("false" | "off"), _ ->
         [(key, "off")]
-      else (
-        debug "Invalid value for ethtool-%s = %s. Must be on|true|off|false."
-          key value ;
+    | Some value, _ ->
+        warn "Invalid value for ethtool-%s = %s. Must be on|true|off|false." key
+          value ;
+        []
+    | None, Some value ->
+        [(key, value)]
+    | None, None ->
         []
-      )
-    else if List.mem_assoc key properties then
-      [(key, List.assoc key properties)]
-    else
-      []
   in
   let speed =
-    if List.mem_assoc "ethtool-speed" oc then
-      let value = List.assoc "ethtool-speed" oc in
-      if value = "10" || value = "100" || value = "1000" then
+    let is_positive_int x =
+      try Scanf.sscanf x "%d%!" (fun i -> i > 0)
```

Comment:
This is clever because `int_of_string` recognizes `0x0d`, for example, but "%d" does not.